lib/internal/modules/esm/shared_constants.js

```diff
@@ -0,0 +1,55 @@
+// This file contains the definition for the constant values that must be
+// available to both the main thread and the loader thread.
+
+'use strict';
+
+/*
+The shared memory area is divided in 4 32-bit long sections. It has to be 32-bit
+long as `Atomics.notify` only works with `Int32Array` objects.
+
+--32-bits-- --32-bits-- --32-bits-- --32-bits--
+    ^           ^           ^           ^
+    |           |           |           |
+    |           |           | NUMBER_OF_INCOMING_ASYNC_RESPONSES
+    |           |           |
+    |           | NUMBER_OF_INCOMING_MESSAGES
+    |           |
+    |  IDLENESS_OF_THREADS
+    |
+WORKER_TO_MAIN_THREAD_NOTIFICATION
+
+WORKER_TO_MAIN_THREAD_NOTIFICATION is only used to send notifications, its value should always be 0.
+
+IDLENESS_OF_THREADS does not represent a number, but an array of booleans.
+Currently, two flags are used: WORKER_THREAD_IS_IDLE, and MAIN_THREAD_IS_IDLE.
+For our use case, a thread is considered IDLE if it has nothing left on its
+event loop and is waiting for the other thread to either terminate the process
+or send it more work.
+If we represent this memory area as a binary number, here's how to interpret its value:
+
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx00: WORKER_THREAD_IS_IDLE is false, and MAIN_THREAD_IS_IDLE is false.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx01: WORKER_THREAD_IS_IDLE is true, and MAIN_THREAD_IS_IDLE is false.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx10: WORKER_THREAD_IS_IDLE is false, and MAIN_THREAD_IS_IDLE is true.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx11: WORKER_THREAD_IS_IDLE is true, and MAIN_THREAD_IS_IDLE is true.
+
+NUMBER_OF_INCOMING_MESSAGES is used to communicate the number of messages sent
+by the main thread and not yet handled by the worker thread. It is also used to
+wake up the worker thread when it locks itself.
```

Comment:
Might we call it a queue, or is that too loaded?

Reply:
What queue are you referring? The 32-bit chunk of memory is only representing a number, it wouldn’t be correct to call that a queue (or I don’t understand what you are suggesting)

Reply:
I meant it represents the count of items in the "queue" (not the "queue" itself).

Reply:
Are you suggesting to rename it to `SIZE_OF_INCOMING_MESSAGE_QUEUE`? I don't think it's much better, but no strong feelings.

Reply:
I was thinking something like

```suggestion
NUMBER_OF_PENDING_MESSAGES is used to track the count of messages awaiting
reply from the worker thread. It is also used to wake up the worker thread when
it locks itself.
```

(abandoned the whole "queue" thing)

Reply:
But it's not at all the number of messages awaiting reply from the worker thread, it's the number of messages that have been sent by the main thread and not yet received by the worker thread.

Reply:
Ahh, I missed that completely. Maybe `NUMBER_OF_MESSAGES_IN_TRANSIT` ("transiting" sounds weird). and then maybe

```suggestion
NUMBER_OF_MESSAGES_IN_TRANSIT is used to communicate the number of messages
sent by the main thread and not yet acknowledged by the worker thread. It is
also used to wake up the worker thread when it locks itself.
```

The "handled" threw me off.